Gems/Terrain/Code/Source/TerrainRenderer/ClipmapBounds.cpp

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <TerrainRenderer/ClipmapBounds.h>
+#include <AzCore/Casting/numeric_cast.h>
+#include <AzCore/Math/MathUtils.h>
+
+namespace Terrain
+{
+    bool ClipmapBoundsRegion::operator==(const ClipmapBoundsRegion& other) const
+    {
+        return m_localAabb == other.m_localAabb && m_worldAabb.IsClose(other.m_worldAabb);
+    }
+    
+    bool ClipmapBoundsRegion::operator!=(const ClipmapBoundsRegion& other) const
+    {
+        return !(*this == other);
+    }
+
+    ClipmapBounds::ClipmapBounds(const ClipmapBoundsDescriptor& desc)
+        : m_size(desc.m_size)
+        , m_halfSize(desc.m_size >> 1)
+        , m_clipmapUpdateMultiple(AZ::GetMax<uint32_t>(desc.m_clipmapUpdateMultiple, 1))
+        , m_scale(desc.m_clipToWorldScale)
+        , m_rcpScale(1.0f / desc.m_clipToWorldScale)
+    {
+        AZ_Error("ClipmapBounds", m_scale > 0.0f, "ClipmapBounds should have a scale that is greater than 0.0f.");
+        m_scale = AZ::GetMax(m_scale, AZ::Constants::FloatEpsilon);
+
+        // recalculate m_center
+        m_center = GetSnappedCenter(GetClipSpaceVector(desc.m_worldSpaceCenter));
+    }
+    
+    auto ClipmapBounds::UpdateCenter(const AZ::Vector2& newCenter, AZ::Aabb* untouchedRegion) -> ClipmapBoundsRegionList
+    {
+        return UpdateCenter(GetClipSpaceVector(newCenter), untouchedRegion);
+    }
+
+    auto ClipmapBounds::UpdateCenter(const Vector2i& newCenter, AZ::Aabb* untouchedRegion) -> ClipmapBoundsRegionList
+    {
+        AZStd::vector<Aabb2i> updateRegions;
+
+        // If the new snapped center isn't the same as the old, then generate update regions in clipmap space
+        Vector2i updatedCenter = GetSnappedCenter(newCenter);
+
+        int32_t xDiff = updatedCenter.m_x - m_center.m_x;
+        int32_t updateWidth = AZStd::GetMin<uint32_t>(abs(xDiff), m_size);
+
+        if (updatedCenter.m_x != m_center.m_x)
```

Comment:
Suggestion: After reading this a couple times, I'm thinking some ASCII art in a comment here would help too:
```
/* Calculate the update regions. In the common case, there will be two update regions that form either an L or inverted L shape. To avoid double-counting the corner, we always put it in the vertical box:
 _ 
| |__
|_|__|
*/
```

Then the two if statements could be commented as "calculate the vertical update box" and "calculate the horizontal update box" (or something like that).